### PR TITLE
Pre commit hook, which was checking unsaved scenes, is not checking also other GameObjects states, like ScriptableObjects

### PR DESCRIPTION
**Pull request by van800**
Pre commit hook, which was checking unsaved scenes, is not checking also other GameObjects states, like ScriptableObjects.

### CODE REVIEW

File: unity/EditorPlugin/AfterUnity56/UnitTesting/Initialization.cs [hunk 1]
```diff
@@ -4,55 +4,115 @@
...
             {
-                if (SceneManager.GetSceneAt(i).isDirty)
-                    return true;
+                var methodInfo = typeof(EditorUtility).GetMethod("IsDirty",
+                    System.Reflection.BindingFlags.Static 
+                    | System.Reflection.BindingFlags.Public 
+                    | System.Reflection.BindingFlags.NonPublic, null, new[] { typeof(int)}, null);
+                if (methodInfo == null)
+                {
+                    ourLogger.Error("IsDirty method not found of type='{0}'", typeof(EditorUtility));
```
Comment (citizenmatt) [inline unity/EditorPlugin/AfterUnity56/UnitTesting/Initialization.cs:103]: Will this show in the Unity Console? Do we need it to be an error? We'll be checking this by default, so for (very) old Unity versions, we'd be showing an error that we can't do anything about.

Comment (van800) [reply unity/EditorPlugin/AfterUnity56/UnitTesting/Initialization.cs:103]: It is an error in the log file, not in the Unity Console.

File: unity/EditorPlugin/AfterUnity56/UnitTesting/Initialization.cs [hunk 2]
```diff
@@ -4,55 +4,115 @@
...
+    public class SpawnManagerScriptableObject : ScriptableObject
     {
-        modelAndLifetime.Model.HasUnsavedScenes.Set(rdVoid => 
+        public string prefabName;
+    
+        public int numberOfPrefabsToCreate;
+        public Vector3[] spawnPoints;
+    }
+                 */
+                var hasDirtyUserAssets = Resources.FindObjectsOfTypeAll<Object>()
```
[the remark this answers is not in the document]
Comment (van800) [reply unity/EditorPlugin/AfterUnity56/UnitTesting/Initialization.cs]: IDK exactly, but only Resources.FindObjectsOfTypeAll works for my use-case.

File: rider/src/main/kotlin/com/jetbrains/rider/plugins/unity/ui/vcs/UnsavedSceneCheckinHandlerFactory.kt [hunk 3]
```diff
@@ -32,25 +32,25 @@ private class UnresolvedMergeCheckHandler(
...
         executor: CommitExecutor?,
         additionalDataConsumer: PairConsumer<Any, Any>
     ): ReturnResult {
-        if (settings.checkUnsavedScenes && project.isUnityProject()) {
+        if (settings.checkUnsavedState && project.isUnityProject()) {
             var providerResult = false
             try {
-                providerResult = project.solution.frontendBackendModel.hasUnsavedScenes
+                providerResult = project.solution.frontendBackendModel.hasUnsavedState
                     .sync(Unit, RpcTimeouts(200L, 200L))
```
[the remark this answers is not in the document]
Comment (van800) [reply rider/src/main/kotlin/com/jetbrains/rider/plugins/unity/ui/vcs/UnsavedSceneCheckinHandlerFactory.kt:54]: I have tested it by generating 7000 assets with ScriptableObject and the call is so fast, that it is unrecognizable in the profiler.
When I tried to create 70000 assets, Unity got stuck processing them.

File: unity/EditorPlugin/AfterUnity56/Initialization.cs [hunk 4]
```diff
@@ -71,7 +71,7 @@ public class SpawnManagerScriptableObject : ScriptableObject
         public Vector3[] spawnPoints;
     }
                  */
-                var hasDirtyUserAssets = Resources.FindObjectsOfTypeAll<Object>()
+                var hasDirtyUserAssets = Resources.FindObjectsOfTypeAll<ScriptableObject>()
```
Comment (citizenmatt) [inline unity/EditorPlugin/AfterUnity56/Initialization.cs:79]: Does this check only apply to `ScriptableObject`? I guess other objects should belong to the scene. What about prefabs?

File: unity/EditorPlugin/AfterUnity56/Initialization.cs [hunk 5]
```diff
@@ -77,14 +77,13 @@ public class SpawnManagerScriptableObject : ScriptableObject
...
-                        if (string.IsNullOrEmpty(assetPath))
+                        if (a.hideFlags < HideFlags.DontSaveInEditor) // this is faster than checking the File.Exists(Path.GetFullPath(assetPath))
                             return false;
-                        
+
                         if (!IsDirty(a))
                             return false;
 
-                        return File.Exists(Path.GetFullPath(assetPath));
+                        return true;
```
Comment (citizenmatt) [inline unity/EditorPlugin/AfterUnity56/Initialization.cs]: Could simplify this to `return IsDirty(a);` but that's a really Minor nitpick

**Review by citizenmatt — APPROVED** [review]
LGTM

File: unity/EditorPlugin/AfterUnity56/Initialization.cs (hunk 4 above)
Comment (van800) [reply unity/EditorPlugin/AfterUnity56/Initialization.cs:79]: I made some experiments and it showed that prefabs are auto-saved.

**Comment by van800** [comment]
merged https://buildserver.labs.intellij.net/buildConfiguration/DotNetDiv_Wave223_TriggerPullRequestMergeIntoNet223/201463110?buildTab=changes&showFiles=false&expandRevisionsSection=true

**Pull request closed without merging**